### PR TITLE
Group subscriber v2

### LINKED ISSUE #336: Create group_subscriber_v2 behavior
Current `group_subscriber` behavior has shortcomings, namely, messages from all partitions end up in a single process. An improved version of this behavior should spawn worker processes for each assigned partition.

### PR DESCRIPTION
**Pull request by k32**
Fixes #336

### CODE REVIEW

File: rebar.config.script [hunk 1]
```diff
@@ -2,6 +2,7 @@ IsRebar3 = erlang:function_exported(rebar3, main, 1),
 DocoptUrl = "https://github.com/zmstone/docopt-erl.git",
 DocOptTag = "0.1.3",
 DocoptDep = {docopt, {git, DocoptUrl, {branch, DocOptTag}}},
+Snabbkaffe = {snabbkaffe, {git, "https://github.com/klarna/snabbkaffe/", {branch, "implement_block_until"}}}, % TODO: master
```
Comment (k32) [inline rebar.config.script]: TODO: merge and release this: https://github.com/klarna/snabbkaffe/pull/14

File: src/brod_group_subscriber_v2.erl [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+  gen_server:call(Pid, {get_committed_offsets, TopicPartitions}, infinity).
+
+%% @doc This function is called only when `partition_assignment_strategy'
+%% is set for `callback_implemented' in group config.
+%% @end
+-spec assign_partitions(pid(), [brod:group_member()],
+                        [{brod:topic(), brod:partition()}]) ->
+        [{member_id(), [brod:partition_assignment()]}].
+assign_partitions(Pid, Members, TopicPartitionList) ->
+  %% TODO: Not implemented
```
Comment (k32) [inline src/brod_group_subscriber_v2.erl]: TODO: Implement this call

**Comment by k32** [comment]
@zmstone Added an implementation sketch of multithreaded group subscriber. Would you mind joining the review?

File: src/brod_group_subscriber_v2.erl [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+  {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling all non call/cast messages
+%% @end
+%%--------------------------------------------------------------------
+-spec handle_info(Info :: timeout() | term(), State :: term()) ->
+                     {noreply, NewState :: term()} |
```
Comment (k32) [inline src/brod_group_subscriber_v2.erl]: TODO: remove autogenerated spec

File: src/brod_group_subscriber_v2.erl [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+%%   The type of message that is going to be handled by the callback
+%%   module. Can be either message or message set.
+%% CbModule:
+%%   Callback module which should have the callback functions
+%%   implemented for message processing.
+%% CbInitArg:
+%%   The term() that is going to be passed to CbModule:init/1 when
+%%   initializing the subscriger.
+%% @end
+-spec start_link(subscriber_config()) -> {ok, pid()} | {error, any()}.
```
Comment (jesperes) [inline src/brod_group_subscriber_v2.erl:168]: Maybe move these docs to the `subscriber_config` type, since they aren't parameters to the function.

**Review by jesperes — APPROVED** [review]

**Comment by zmstone** [comment]
very nice! will take a closer look later.

File: src/brod_group_subscriber_v2.erl [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+  {ok, State}.
+
+handle_call({get_committed_offsets, TopicPartitions}, _From,
+            #state{ cb_module = CbModule
+                  , workers   = Workers
+                  } = State) ->
+  Fun = fun(TopicPartition) ->
+            case Workers of
+              #{TopicPartition := Pid} ->
+                case CbModule:get_committed_offset(Pid) of
```
Comment (k32) [inline src/brod_group_subscriber_v2.erl]: This is wrong. Callback expects state.

**Review by helanhalvan — APPROVED** [review]

**Comment by qzhuyan** [comment]
As talked in person, just mark two things for next PR:
- not to have 1 worker per partition. One worker process should work for multiple partitions of same topic that would reduce context switching costs.
- It is risky to have infinity timeout in api calls. should intro some timeouts.

**Review by helanhalvan — APPROVED** [review]

**Comment by vikas15bhardwaj** [comment]
Hi All, when is your plan to release group_subscriber_v2. I am working on an elixir library and having some issues using current 3.8.1 group_subscriber (not able to have two group subscriber in parallel for same topic). It looks like v2 version would fix it.

**Comment by k32** [comment]
Hi, @vikas15bhardwaj ! I am planning to release this feature as soon as I stress-test it on our internal app (1-2 weeks?). _v2 hasn't been used for anything serious yet, so I don't want to rush the release. API won't change though, so you can try to experiment with the `master`. Please don't hesitate to report any problems that you encounter, I am highly interested in this.

Also please keep in mind that #347 can bite you if you have too many partitions. In its current form _v2 doesn't limit the number of processes.

**Comment by vikas15bhardwaj** [comment]
We do have serious use case for using it and it will surely help to get this version to test at our end. I'll see if I am able to work with master branch directly.

**Comment by k32** [comment]
Hi @vikas15bhardwaj ! Apologies for being so slow with brod releases. I just published version 3.9.0 that includes the _v2 behavior. Please note that I still consider it experimental.

**Pull request merged**